### PR TITLE
Basic implementation for `basicUnsafeMove`

### PR DESCRIPTION
**Pull request by dmalkr**
Implementation for `basicUnsafeMove` and my own test for this -- it just goes through all the offset/stride/length options of source/target vectors. Arbitrary tests from `vector` package still not adopted.

Please review.

### CODE REVIEW

File: vecvec-lapack/Vecvec/LAPACK/Internal/Compat.hs [hunk 1]
```diff
@@ -33,3 +36,13 @@ updPtr f (ForeignPtr p c) = case f (Ptr p) of { Ptr q -> ForeignPtr q c }
...
 #endif
+
+-- | Distance in elements between two pointers. Supplement for `advancePtr`.
+--
+distancePtr :: forall a . Storable a => Ptr a -> Ptr a -> Int
+-- distancePtr ptrFrom ptrTo = (ptrTo `minusPtr` ptrFrom) `div` (sizeOf (undefined :: a))
+distancePtr = distancePtr' undefined -- TODO WTF?? why ambigous??
+ where
+    distancePtr' :: Storable a => a -> Ptr a -> Ptr a -> Int
+    distancePtr' x ptrFrom ptrTo = (ptrTo `minusPtr` ptrFrom) `div` sizeOf x
```
Comment (Shimuuar) [inline vecvec-lapack/Vecvec/LAPACK/Internal/Compat.hs]: Without `ScopedTypeVariables` GHC will think that `a` is some unrelated type variable.

File: vecvec-lapack/tests/TST/Memory.hs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import           Vecvec.Classes.NDArray                (Length (..), slice)
+import qualified Vecvec.LAPACK                         as VV
+import           Vecvec.LAPACK.Internal.Vector.Mutable (Strided (..))
+
+import           Test.Tasty
+import           Test.Tasty.HUnit
+
+
+tests :: TestTree
+tests = testGroup "memory"
```
Comment (Shimuuar) [inline vecvec-lapack/tests/TST/Memory.hs:18]: Test could be simplified. Specification for move is quite simple: If vectors overlap, the copying is performed as if the source vector were copied to a temporary vector and then the temporary vector was copied to the target vector. So two program should be equivalent:

```haskell
carrier <- MVG.generateM carrierLength id
let source =  slice ... carrier
let target =  slice ... carrier
MVG.move target source
VG.unsafeFreeze carrier
```

```haskell
carrier <- MVG.generateM carrierLength id
let source =  slice ... carrier
let target =  slice ... carrier
buf <- MVG.clone source
MVG.copy target buf
VG.unsafeFreeze carrier
```

File: vecvec-lapack/Vecvec/LAPACK/Internal/Vector/Mutable.hs [hunk 3]
```diff
@@ -212,17 +212,59 @@ instance VS.Storable a => MVG.MVector MVec a where
...
+    let intersectPosition = diff `div` strideDiff
+    if diff /= 0 && strideDiff /= 0 && intersectPosition >= 0 && intersectPosition < len then do
+      if fpTarget < fpSource then do
+        copy 0         (intersectPosition + 1)
+        copy (len - 1) intersectPosition
+      else do
+        copy intersectPosition (-1)
+        copy (intersectPosition + 1) len
+    else
+      case () of
```
Comment (Shimuuar) [inline vecvec-lapack/Vecvec/LAPACK/Internal/Vector/Mutable.hs]: This is purely syntactic. `if ... then ... else case () of ...` could be replaced with `MultiWayIf`

**Review by Shimuuar — COMMENTED** [review]
Overall  everything looks good. Except test could be simplified

File: vecvec-lapack/Vecvec/LAPACK/Internal/Compat.hs (hunk 1 above)
Comment (dmalkr) [reply vecvec-lapack/Vecvec/LAPACK/Internal/Compat.hs]: Weird. I thought I tried `ScopedTypeVariables` :)

File: vecvec-lapack/tests/TST/Memory.hs (hunk 2 above)
Comment (dmalkr) [reply vecvec-lapack/tests/TST/Memory.hs:18]: Done

File: vecvec-lapack/Vecvec/LAPACK/Internal/Vector/Mutable.hs (hunk 3 above)
Comment (dmalkr) [reply vecvec-lapack/Vecvec/LAPACK/Internal/Vector/Mutable.hs]: Done

File: vecvec-lapack/tests/TST/Memory.hs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                      let target1 =  slice (Strided (targetOffset, Length (targetLength * targetStride)) targetStride) carrier1
+                      MVG.basicUnsafeMove target1 source1
+                      --
+                      carrier2    <- MVG.generateM carrierLength pure
+                      let source2 =  slice (Strided (sourceOffset, Length (sourceLength * sourceStride)) sourceStride) carrier2
+                      let target2 =  slice (Strided (targetOffset, Length (targetLength * targetStride)) targetStride) carrier2
+                      tmp <- MVG.clone source2
+                      MVG.basicUnsafeCopy target2 tmp
+                      --
+                      (,,,) <$> VG.unsafeFreeze target1 <*> VG.unsafeFreeze target2 <*> VG.unsafeFreeze carrier1 <*> VG.unsafeFreeze carrier2
```
Comment (dmalkr) [inline vecvec-lapack/tests/TST/Memory.hs]: This use of the applicative and `unsafeFreeze` is safe, right?

**Pull request merged**